### PR TITLE
Replace offsets with iterators in cuIO utilities and CSV parser

### LINKED ISSUE #6210: [FEA] cuio: reduce subscript member access in parse utilities
There's a lot of clutter cause by subscript member access in the parsing utilities which inhibits refactoring. Instead of subscript member access, we can use pointers as iterators. Updating return values to use pointers rather than offsets will also simplify the functions. Invocations will need to be adjusted to account for the removal of a sentinel value (`-1`) in place of `end` in some locations.

Example:
https://github.com/rapidsai/cudf/blob/c27d844986b70003c16b6203464f53936a41739c/cpp/src/io/csv/datetime.cuh#L32-L42

Raw replacement:
```c++
__inline__ __device__ const char *begin findFirstOccurrence(const char *begin, const char *end, char c)
{
  while (begin < end and *begin != c) { begin++; }
  return begin;
} 
```

This specific example is probably just `thrust::find`, but there are numerous others that may be more complex.

### PR DESCRIPTION
**Pull request by vuule**
Closes #6210 

- Rename datetime utils to snake_case;
- Rename datetime utils so that `parse_xyz` functions move the input iterator past the parsed value, and `to_xyz` function do not change the input iterators.
- Replace `findFirstOccurrence` with `thrust::find`.
- Replace use of offsets with pointers in `datetime.cuh` and `csv_gpu.cu`;
- Rename some variables in the CSV parser to make the code clearer.

Note: the semantics of variables/parameters did not change in this PR - `T* end` points to the last element in the range in many places.

### CODE REVIEW

**Review by harrism — APPROVED** [review]
Good improvement!

File: cpp/src/io/csv/datetime.cuh [hunk 1]
```diff
@@ -355,96 +315,136 @@ __inline__ __device__ int64_t parseDateTimeFormat(char const* begin, char const*
...
   return answer;
 }
 
-// parse integer and update the start position
+/**
+ * @brief Parses the input string into an integral value of the given type.
+ *
+ * Moves the `begin` iterator past the parsed value.
+ *
+ * @param begin Pointer to the first element of the string
```
Comment (cwharris) [inline cpp/src/io/csv/datetime.cuh:336]:
```suggestion
 * @param[in, out] begin Pointer to the first element of the string
```

File: cpp/src/io/csv/datetime.cuh [hunk 2]
```diff
@@ -355,96 +315,136 @@ __inline__ __device__ int64_t parseDateTimeFormat(char const* begin, char const*
...
+  //:%M:%S
+  if (*cur == sep) {
+    ++cur;
+    minute = parse_integer<int8_t>(&cur, end);
+  }
+  int8_t second{0};
+  if (*cur == sep) {
+    ++cur;
+    second = parse_integer<int8_t>(&cur, end);
+  }
```
Comment (cwharris) [inline cpp/src/io/csv/datetime.cuh:436]: It'd be cool if we could declare and init simultaneously, but this isn't super clean either. :/
```suggestion
  //:%M:%S
  auto minute = parse_optional_integer<int8_t>(&cur, end, sep);
  auto second = parse_optional_integer<int8_t>(&cur, end, sep);
```
```c++
template<typename T>
parse_optional_integer(char** cur, char* end, char sep) {
  if (**cur == sep) {
    (*cur)++;
    return parse_integer<T>(cur, end);
  }
  return 0;
}
```

**Review by cwharris — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Pull request merged**